### PR TITLE
[2/?] Instant loop out:  Add reservations

### PR DESCRIPTION
**Pull request by sputn1ck**
Prior PR: https://github.com/lightninglabs/loop/pull/631

**Skip 1st commit**

This PR adds reservation functionality to loop.

Reservations are 2-of-2 musig2 taproot outputs with a single funder. The output contains an expiry path that allows the funder to spend the funds after N blocks.

Reservations can be used a new type of loop out, where the server commits funds to the 2-of-2 output and the loop client can then trustlessly and instantaneously loop out, allow for rapid rebalancing of channels.

The reservation state machine in this PR looks like this:

```mermaid
stateDiagram-v2
[*] --> Init: OnServerRequest
Confirmed
Confirmed --> TimedOut: OnTimedOut
Confirmed --> Confirmed: OnRecover
Failed
Init
Init --> Failed: OnError
Init --> WaitForConfirmation: OnBroadcast
Init --> Failed: OnRecover
TimedOut
WaitForConfirmation
WaitForConfirmation --> WaitForConfirmation: OnRecover
WaitForConfirmation --> Confirmed: OnConfirmed
WaitForConfirmation --> TimedOut: OnTimedOut
```

TODO:
- [x] SQL db
- [x] Recover reservation fsms on startup
- [x] More Unit test coverage

Next PR:  https://github.com/lightninglabs/loop/pull/633

### CODE REVIEW

**Review by hieblmi — COMMENTED** [review]
This is great work @sputn1ck. I think future features will benefit from your code structure. I've just left a couple of things that sprung to my eye so far.

**Review by hieblmi — COMMENTED** [review]
Some more comments.

**Review by hieblmi — COMMENTED** [review]
Some more comments

File: fsm/fsm.go
Comment (hieblmi) [inline fsm/fsm.go:81]: nit: could you create a separate PR for `fsm/fsm.go` and `fsm/observer.go` changes?

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:43]: nit: log with capital.

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:52]: nit: could use `r.ctx`.

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:68]: Can it cause any issues that we set `r.reservation` before ensuring that we were able to store it?

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:91]: nit: capitalize (won't comment this elsewhere, but applies to all logs).

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:92]: nit: missing `:`

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:134]: nit: I guess it's less of an issue here but should we always set members after ensuring that the DB state is updated?

File: reservation/actions.go
Comment (bhandras) [inline reservation/actions.go:151]: nit: iiuc it's not a TODO right? We can just return `NoOp` here.

File: reservation/interfaces.go
Comment (bhandras) [inline reservation/interfaces.go:62]: nit: missing godocs for exported fields.

File: reservation/interfaces.go
Comment (bhandras) [inline reservation/interfaces.go:79]: nit: missing godoc.

File: reservation/interfaces.go
Comment (bhandras) [inline reservation/interfaces.go:84]: Alternatively could add a `NewReservation` function that sets these fields and does the prechecks.

File: loopdb/sqlc/migrations/000003_reservations.up.sql
Comment (bhandras) [inline loopdb/sqlc/migrations/000003_reservations.up.sql:1]: nit: please add some comments.

File: reservation/store.go
Comment (bhandras) [inline reservation/store.go:112]: q: This is UTC?

File: reservation/manager.go
Comment (bhandras) [inline reservation/manager.go:148]: nit: missing godoc.

File: reservation/manager.go
Comment (bhandras) [inline reservation/manager.go:178]: Shouldn't this goroutine (and this whole function) use the `m.runCtx` instead?

File: reservation/manager.go
Comment (bhandras) [inline reservation/manager.go:58]: Just to be sure I'd move this to below where we set the run ctx.

File: reservation/manager.go
Comment (bhandras) [inline reservation/manager.go:235]: What's the benefit of checking these members (which should be set at this point) vs just letting the program crash instead?

File: loopd/daemon.go
Comment (bhandras) [inline loopd/daemon.go:596]: nit: defer log instead.

**Review by bhandras — COMMENTED** [review]
Excellent work, I added some comments, most of them are nits. Almost ready!

**Review by hieblmi — COMMENTED** [review]
Awesome work @sputn1ck, this is looking really good. I've left more code design and style nits this time and also have some clarifying questions.

File: reservation/actions.go
Comment (hieblmi) [reply reservation/actions.go:91]: wanna have `pkscript: %v, initiationheight: %v`

File: reservation/actions.go
Comment (hieblmi) [reply reservation/actions.go:68]: Sorry my bad, I thought that `CreateReservation` has access to the `r.reservation`, so we could've omitted `r.ctx, reservation` in the method sig.

File: instantout/reservation/fsm.go
Comment (bhandras) [inline instantout/reservation/fsm.go:74]: nit: I think these states deserve some godocs too. Just some high level explanation what they mean (also the events).

File: instantout/reservation/actions_test.go
Comment (bhandras) [inline instantout/reservation/actions_test.go:90]: nit: missing godoc.

File: loopdb/sqlc/queries/reservations.sql
Comment (bhandras) [inline loopdb/sqlc/queries/reservations.sql:44]: nit: why not just use `*` (below too).

File: loopdb/sqlc/queries/reservations.sql
Comment (bhandras) [inline loopdb/sqlc/queries/reservations.sql:58]: nit: no need for the `reservations.` prefix (below too).

File: loopdb/sqlc/queries/reservations.sql
Comment (bhandras) [inline loopdb/sqlc/queries/reservations.sql:84]: nit: no need for the prefix.

File: instantout/reservation/store.go
Comment (bhandras) [inline instantout/reservation/store.go:156]: nit: maybe check that there's at least one update. Not strictly needed but better to be safe.

File: instantout/reservation/store.go
Comment (bhandras) [inline instantout/reservation/store.go:198]: nit: similarly could ensure that there's at least one update.

File: instantout/reservation/store_test.go
Comment (bhandras) [inline instantout/reservation/store_test.go:16]: nit: missing godoc.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:29]: nit: there's no need for the `runCtx` as the context created inside `Run` could just be passed to all other functions.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:51]: nit: i think we can also avoid storing the current height, could just be a local variable and could be passed to `newReservation` so no need for the mutex either.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:181]: nit: to cut complexity a bit we could instead check `if err == nil` and then we don't need this nesting below.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:208]: Instead of sleeping would be nicer to use `time.After` and (initialize it with zero so it fires immediately), that way terminating would be a non-issue.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:243]: q: Do we need this goroutine at all?

File: loopd/daemon.go
Comment (bhandras) [inline loopd/daemon.go:582]: This is to ensure that LND is initialized? Maybe add a comment to explain.

File: cmd/loop/reservations.go
Comment (bhandras) [inline cmd/loop/reservations.go:14]: nit: A bit more help could be useful so users understand what the reservations are etc.

**Review by bhandras — APPROVED** [review]
LGTM (pending the last remaining nits). Awesome work, great to see this getting into ready state!!! 🎉 🎉

File: instantout/reservation/interfaces.go
Comment (hieblmi) [inline instantout/reservation/interfaces.go:49]: maybe better to have these magic nums behind variables

File: loopdb/sqlc/migrations/000003_reservations.up.sql
Comment (hieblmi) [inline loopdb/sqlc/migrations/000003_reservations.up.sql:43]: This is dropped as `reservations_updates`.

**Review by hieblmi — CHANGES_REQUESTED** [review]
Looking good, I have some nits left and two things we should address before approval.

1.) We should decide on the key family for client keys, maybe put them to where loop currently has it `swap/keychain.go`.
2.) The reservation update table is dropped with an invalid name.

File: loopd/daemon.go
Comment (bhandras) [reply loopd/daemon.go:584]: Cool 👍

File: instantout/reservation/manager.go
Comment (bhandras) [reply instantout/reservation/manager.go:242]: 👍

File: instantout/reservation/store.go
Comment (bhandras) [inline instantout/reservation/store.go:57]: nit: a nice thing to have is to inject a clock as a dependency, so we can set  deterministic expectations in tests.

File: instantout/reservation/manager.go
Comment (bhandras) [inline instantout/reservation/manager.go:168]: I think we don't need this check here, since the stream will fail if the context is canceled anyway. And also we below check in the reconnect loop too.

**Review by bhandras — APPROVED** [review]
LGTM 💯

File: instantout/reservation/manager.go
Comment (hieblmi) [inline instantout/reservation/manager.go:107]: nit: incomplete comment

**Review by hieblmi — APPROVED** [review]
Great work @sputn1ck!

File: instantout/reservation/actions_test.go
Comment (bhandras) [inline instantout/reservation/actions_test.go:226]: nit: comment is a bit off.

**Review by bhandras — APPROVED** [review]
Re-reviewed due to added test coverage. LGTM 🎉 🎉

**Pull request merged**
